src/Data/Serialize/Get.hs

```diff
@@ -402,30 +418,101 @@ ensure n0 = n0 `seq` Get $ \ s0 b0 m0 w0 kf ks -> let
+
+isolationUnderParse :: Get a
+isolationUnderParse =  fail "Isolated parser didn't consume all input"
+
+isolationUnderSupply :: Get a
+isolationUnderSupply = fail "Too few bytes supplied to isolated parser"
+
+isolate0 :: Get a -> Get a
+isolate0 parser = do
+  rest <- get
+  cur <- bytesRead
+  put mempty cur
+  a    <- parser
+  put rest cur
+  pure a
+
 -- | Isolate an action to operating within a fixed block of bytes.  The action
 --   is required to consume all the bytes that it is isolated to.
 isolate :: Int -> Get a -> Get a
-isolate n m = do
-  M.when (n < 0) (fail "Attempted to isolate a negative number of bytes")
-  s <- ensure n
-  let (s',rest) = B.splitAt n s
-  cur <- bytesRead
-  put s' cur
-  a    <- m
-  used <- get
-  unless (B.null used) (fail "not all bytes parsed in isolate")
-  put rest (cur + n)
-  return a
+isolate n m
+  | n < 0 = negativeIsolation
+  | n == 0 = isolate0 m
+  | otherwise = do
+      s <- ensure' n
+      let (s',rest) = B.splitAt n s
+      cur <- bytesRead
+      put s' cur
+      a    <- m
+      used <- get
+      unless (B.null used) isolationUnderParse
+      put rest (cur + n)
+      return a
+
+getAtMost :: Int -> Get B.ByteString
+getAtMost 0 = pure mempty
+getAtMost n = do
+  (bs, rest) <- B.splitAt n <$> ensure' 1
+  curr <- bytesRead
+  put rest (curr + B.length bs)
+  pure bs
+
+-- | An incremental version of 'isolate', which doesn't try to read the input
+--   into a buffer all at once.
+isolateLazy :: Int -> Get a -> Get a
+isolateLazy n parser
+  | n < 0 = negativeIsolation
+  | n == 0 = isolate0 parser
+isolateLazy n parser = do
+  initialBytesRead <- bytesRead
+  bs <- getAtMost n
+  go initialBytesRead $ runGetPartial parser bs
+  where
+    go :: Int -> Result a -> Get a
+    go initialBytesRead r = case r of
+      FailRaw (msg, stack) bs -> do
+        m <- bytesRead
+        put bs m
+        failRaw msg stack
+      Done a bs -> do
+        bytesRead' <- bytesRead
+        unless (bytesRead' - initialBytesRead == n && B.null bs) isolationUnderParse
+        pure a
+      Partial cont -> do
+        pos <- bytesRead
+        -- Sometimes, the amount of bytes left in the isolated stream
+        -- will be zero (ie. `n - (pos - initialBytesRead) = 0`)
+        -- In this case we give the current continuation an empty
+        -- bytestring. We do this so that the inner parser can determine
+        -- the final result.
+        -- If, however, the inner parser returns another continuation,
+        -- then presumably it will keep doing so, for as long as we provide
+        -- it with empty chunks of input.
+        -- In this case, we throw our own error (`isolationUnderSupply`),
+        -- to avoid recurring indefinitely.
+        bs <- getAtMost $ n - (pos - initialBytesRead)
+        if B.null bs
+          then case cont bs of
+            Partial _ -> isolationUnderSupply
+            a -> go initialBytesRead a
+          else go initialBytesRead $ cont bs
```

Comment:
Thinking more about this, I'd really like it if this case didn't give the partial result one last chance with an empty buffer. My reasoning is that: the partial case should only happen if a non-zero input was requested, and not enough bytes were available, so the fact that a parser could succeed here means that it requested zero bytes. Requesting zero bytes should always immediately succeed as we can always construct an empty bytestring, so it should be impossible to get to this case.

Would you be okay with replacing this logic with a guard that there is space remaining in the isolate budget, and then continuing only when the buffer is non-empty?

Reply:
Sure thing!